docs/manpages/mdbook-manmunge/src/preprocess.rs

```diff
@@ -0,0 +1,225 @@
+use std::io;
+
+use mdbook::book::{Book, Chapter};
+use mdbook::preprocess::{CmdPreprocessor, Preprocessor, PreprocessorContext};
+use mdbook::BookItem;
+use pulldown_cmark::{Event, HeadingLevel, Options, Parser, Tag, TagEnd};
+use semver::{Version, VersionReq};
+use url::{ParseError, Url};
+
+use crate::error::Error;
+use crate::verbatim::{Cmark, CmarkTable, Verbatim};
+
+struct MdbookMunge;
+
+impl MdbookMunge {
+    fn new() -> Self {
+        Self
+    }
+}
+
+impl Preprocessor for MdbookMunge {
+    fn name(&self) -> &str {
+        "manmunge"
+    }
+
+    fn run(&self, _ctx: &PreprocessorContext, mut book: Book) -> mdbook::errors::Result<Book> {
+        // We don't have any arguments yet, but if/when we do, they can be extracted from
+        // `ctx.config.get_preprocessor(self.name())`
+
+        book.for_each_mut(|item| {
+            let BookItem::Chapter(chapter) = item else {
+                return;
+            };
+            if chapter.is_draft_chapter() {
+                return;
+            }
+
+            match rewrite_chapter(chapter) {
+                Ok(result) => chapter.content = result,
+                Err(error) => log::error!("{}: Could not process chapter ({error})", chapter.name),
+            }
+        });
+
+        Ok(book)
+    }
+}
+
+fn event_type<'parse>(event: &Event<'parse>) -> &'static str {
+    match event {
+        Event::Start(Tag::List(Some(_))) | Event::End(TagEnd::List(true)) => "ordered list",
+        Event::Start(Tag::List(None)) | Event::End(TagEnd::List(false)) => "unordered list",
+        Event::Start(Tag::Table(_)) | Event::End(TagEnd::Table) => "table",
+        _ => unreachable!(),
+    }
+}
+
+fn rewrite_chapter(chapter: &mut Chapter) -> Result<String, Error> {
+    let mut buf = String::with_capacity(chapter.content.len());
+
+    let mut options = Options::empty();
+    options.insert(Options::ENABLE_TABLES);
+    let parser = Parser::new_ext(&chapter.content, options);
+
+    let mut strip_h1 = false;
+    let mut strip_link = false;
+    let mut verbatim: Option<Box<dyn Verbatim>> = None;
+
+    let events = parser
+        .flat_map(|event| match event {
+            // String links with relative and unparsable URLs
+            Event::Start(Tag::Link { ref dest_url, .. }) => {
+                let url = Url::parse(dest_url);
+
+                match url {
+                    Ok(_) => vec![Some(event)],
+
+                    Err(ParseError::RelativeUrlWithoutBase) => {
+                        log::info!("{}: Stripping relative URL {dest_url}", chapter.name);
```

Comment:
should the strips be debug level? I haven't run this but I'm curious as to the volume of these stripping events.